### PR TITLE
Refactor Redux persistence for better types

### PR DESCRIPTION
**Pull request by jsnajdr**
Implements the plan outlined in https://github.com/Automattic/wp-calypso/pull/49618#discussion_r570980750 🙂 

The `SERIALIZE` and `DESERIALIZE` Redux actions are completely gone. Now the serialization handlers are added as properties on the reducer function:
```js
const reducer = ( state, action ) {
  switch ( action.type ) { ... }
}
reducer.serialize = state => persisted;
reducer.deserialize = persisted => state;
```

There are `serialize` and `deserialize` helpers in `calypso/state/utils` that can handle also plain reducers without these properties and serialize them in the `withoutPersistence` way: serialize to `undefined` and deserialize to initial state. Called as:
```js
const persisted = serialize( reducer, state );
const state = deserialize( reducer, persisted );
```

There is also a new `withPersistence` wrapper that adds serialization to a reducer. Usages:
```js
// Adds identity `serialize` and `deserialize` transforms
const persistingReducer = withPersistence( reducer );

// Adds custom `serialize` and `deserialize` transforms
const persistingReducer = withPersistence( reducer, {
  serialize: state => state.toString(),
  deserialize: persisted => CustomState.fromString( persisted )
} );
```

All we need to do now is to test thoroughly and merge this step-by-step 🙂 There's already one spinoff in #50221.

### CODE REVIEW

**Comment by tyxla** [comment]
Marked as "Needs rebase" since #50223 is another spin-off 😉

**Comment by jsnajdr** [comment]
After a long preparatory work, this PR is now ready for final review. It:
- eliminates the `SERIALIZE` and `DESERIALIZE` Redux action types completely, in favor of calling `.serialize()` and `.deserialize()` methods attached to the reducer function.
- rewrites the `serialize` and `deserialize` helpers to use the new API internally
- rewrites the `withPersistence`, `withoutPersistence`, `withSchemaValidation`, `combineReducers` and `keyedReducers` helpers to do persistence the new way
- updates JSDoc comments and Markdown documentation about persistence

Note that the code (non-doc) changes in this PR are 100% constrained to the `client/state/utils` module and changing its internal implementation. No app code is changed, no unit tests are changed. That preparatory work to abstract all internal implementation details was done in previous spinoff commits. Now we just change the internals, keeping the public API intact.

File: client/state/utils/keyed-reducer.js [hunk 1]
```diff
@@ -100,37 +82,9 @@ export const keyedReducer = ( keyPath, reducer ) => {
 		);
 	}
 
-	const persistingReducer = setupReducerPersistence( reducer );
-
-	const initialState = persistingReducer( undefined, { type: '@@calypso/INIT' } );
+	const initialState = reducer( undefined, { type: '@@calypso/INIT' } );
 
 	const combinedReducer = ( state = {}, action ) => {
-		if ( action.type === SERIALIZE ) {
```
Comment (tyxla) [inline client/state/utils/keyed-reducer.js:108]: Let's also update the inline docs on L30 above to no longer mention the old serialization action types.

File: client/state/utils/schema-utils.js [hunk 2]
```diff
@@ -105,31 +105,26 @@ export const withSchemaValidation = ( schema, reducer ) => {
 		throw new Error( 'null schema passed to withSchemaValidation' );
 	}
 
-	const wrappedReducer = ( state, action ) => {
-		if ( action.type === DESERIALIZE ) {
```
Comment (tyxla) [inline client/state/utils/schema-utils.js:109]: Let's also update the JSdoc above - specifically it contains some examples with the old action types.

**Review by tyxla — COMMENTED** [review]
This looks and works great! Nice work on this effort, Jarda 👏 

There are some occurrences of the old serialization action types in documentation, I think we should fix that. There is also https://github.com/Automattic/wp-calypso/pull/50223#discussion_r600503847 and some occurrences in the `createReducer` removal codemod, which seems obsolete and unnecessary by now and could be removed.

**Comment by jsnajdr** [comment]
@tyxla Review comments addressed.

The `timezone` reducer test is fixed. In this case, the `serialize()` call is just an identity that returns the input state, so any `reducer( state, { type: 'random' } )` call would continue to work, too.

The JSDoc comments still mentioning `SERIALIZE` are updated.

> some occurrences in the `createReducer` removal codemod, which seems obsolete

We tend to keep the codemods in the repo even after they are obsolete. After all, all of them become obsolete after they are used. The idea is that writing codemods is hard, and having some inspirational examples at hand is very useful when writing new ones.

**Review by tyxla — APPROVED** [review]
This is working well and looking great. Ready to go 🚀 

Amazing work on it Jarda 👏

**Pull request merged**